### PR TITLE
[Infrastructure] Add Session Management Support

### PR DESCRIPTION
**Pull request by youssefsoli**
Added session management support to the project, and also added some implementations of it for the existing use cases.

[Show request errors with Spring](https://github.com/CSC207-2022F-UofT/course-project-fitbook/commit/b75c4700858a7fd477f6331e1ee6aa0f8faa4fb3)
  
[Add Spring Session dependency](https://github.com/CSC207-2022F-UofT/course-project-fitbook/commit/359f78c1b352d29c8ecf36c0d31ee4408cf5b287)

[Add unauthorized user exception](https://github.com/CSC207-2022F-UofT/course-project-fitbook/commit/80dc773c5c561ae2a0e953cfc85b55b86a0fa0b4)
  
[Save session on login and register](https://github.com/CSC207-2022F-UofT/course-project-fitbook/commit/95ac91b75a60efec6c854bb667e6f82df2ab9d4a)
  
[Add user logout controller](https://github.com/CSC207-2022F-UofT/course-project-fitbook/commit/b418325db4a151bea463d412773e3e25cf3f0c85)
  
[Use ID in user session for follow use case](https://github.com/CSC207-2022F-UofT/course-project-fitbook/commit/57cd6731264e6a080eda241ec0f2532cd91cd20c)

### CODE REVIEW

**Review by jpablo2002 — APPROVED** [review]
Everything looks good to me! Great job

File: src/main/java/ca/utoronto/fitbook/adapter/web/FollowController.java [hunk 1]
```diff
@@ -9,16 +10,24 @@
...
 {
     private final FollowUseCase followUseCase;
 
     @PostMapping(path = "/follow")
-    String followUser(Model model, @RequestBody FollowCommand command) {
+    String followUser(Model model, HttpSession session, @RequestBody FollowRequestBody body) {
+        String userId = (String) session.getAttribute("userId");
+        if (userId == null)
+            throw new UnauthorizedUserException();
+        FollowCommand command = new FollowCommand(userId, body.getFolloweeId());
```
Comment (modaserMoj) [inline src/main/java/ca/utoronto/fitbook/adapter/web/FollowController.java:26]: Is it supposed to be followee?

File: src/main/java/ca/utoronto/fitbook/adapter/web/FollowRequestBody.java [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package ca.utoronto.fitbook.adapter.web;
+
+import lombok.*;
+
+@Getter
+@Setter
+public class FollowRequestBody
+{
+    @NonNull
+    String followeeId;
```
Comment (modaserMoj) [inline src/main/java/ca/utoronto/fitbook/adapter/web/FollowRequestBody.java:10]: Same as above

File: src/main/java/ca/utoronto/fitbook/adapter/web/FollowController.java (hunk 1 above)
Comment (youssefsoli) [reply src/main/java/ca/utoronto/fitbook/adapter/web/FollowController.java:26]: Yes the followee is the person being followed

**Review by modaserMoj — APPROVED** [review]
Sgtm

File: src/main/java/ca/utoronto/fitbook/adapter/web/UserLoginController.java [hunk 3]
```diff
@@ -1,26 +1,28 @@
...
+
 @Controller
 @RequiredArgsConstructor
 public class UserLoginController {
     private final UserLoginUseCase userLoginUseCase;
 
     @PostMapping(path = "/login")
-    String UserLogin(Model model, @RequestBody UserLoginCommand command) {
+    String loginUser(Model model, HttpSession session, @RequestBody UserLoginCommand command) {
         UserLoginResponse response = userLoginUseCase.loginUser(command);
```
Comment (JacketPants12) [inline src/main/java/ca/utoronto/fitbook/adapter/web/UserLoginController.java:21]: Shoudln't it cehck if userId in session is null like above?

File: src/main/java/ca/utoronto/fitbook/adapter/web/UserRegisterController.java [hunk 4]
```diff
@@ -9,15 +9,20 @@
...
 @Controller
 @RequiredArgsConstructor
 public class UserRegisterController {
     private final UserRegisterUseCase userRegisterUseCase;
 
     @PostMapping(path = "/register")
-    String UserRegister(Model model, @RequestBody UserRegisterCommand command) {
+    String registerUser(Model model, HttpSession session, @RequestBody UserRegisterCommand command) {
         UserRegisterResponse response = userRegisterUseCase.createUser(command);
+        session.setAttribute("userId", response.getId());
```
Comment (JacketPants12) [inline src/main/java/ca/utoronto/fitbook/adapter/web/UserRegisterController.java:22]: check for session userId null?

**Review by JacketPants12 — CHANGES_REQUESTED** [review]
Small clarification and/or fix needed for check of null userIds in session.

File: src/main/java/ca/utoronto/fitbook/adapter/web/UserLoginController.java (hunk 3 above)
Comment (youssefsoli) [reply src/main/java/ca/utoronto/fitbook/adapter/web/UserLoginController.java:21]: No, since the session should be blank on login and register.

**Review by JacketPants12 — APPROVED** [review]
Session management is an important security feature. Therefore the changes made in this pull request are appropriate as all concerns were addressed by the author.

**Pull request merged**